commands/list/nodepools/command.go

```diff
@@ -0,0 +1,206 @@
+	NODES MIN/MAX:  The minimum and maximum number of worker nodes in this pool
+	NODES DESIRED:  Current desired number of nodes as determined by the autoscaler
+	NODES READY:    Number of nodes that are in the Ready state in kubernetes
+	CPUS:           Sum of CPU cores in nodes that are in state Ready
+	RAM (GB):       Sum of memory in GB of all nodes that are in state Ready
+
+To see all available details for a cluster, use 'gsctl show nodepool <cluster-id>/<nodepool-id>'.
+
+To list all clusters you have access to, use 'gsctl list clusters'.`,
+		PreRun: printValidation,
+		Run:    printResult,
+	}
+)
+
+const activityName = "list-nodepools"
+
+type arguments struct {
+	apiEndpoint string
+	authToken   string
+	scheme      string
+	clusterID   string
+}
+
+// defaultArgs creates arguments based on command line flags and config.
+func defaultArgs(cmdLineArgs []string) arguments {
+	endpoint := config.Config.ChooseEndpoint(flags.CmdAPIEndpoint)
+	token := config.Config.ChooseToken(endpoint, flags.CmdToken)
+	scheme := config.Config.ChooseScheme(endpoint, flags.CmdToken)
+
+	return arguments{
+		apiEndpoint: endpoint,
+		authToken:   token,
+		scheme:      scheme,
+		clusterID:   cmdLineArgs[0],
+	}
+}
+
+func verifyPreconditions(args arguments, positionalArgs []string) error {
+	if config.Config.Token == "" && args.authToken == "" {
+		return microerror.Mask(errors.NotLoggedInError)
+	}
+
+	return nil
+}
+
+func printValidation(cmd *cobra.Command, positionalArgs []string) {
+	args := defaultArgs(positionalArgs)
+	err := verifyPreconditions(args, positionalArgs)
+	if err == nil {
+		return
+	}
+
+	errors.HandleCommonErrors(err)
+}
+
+func fetchNodePools(args arguments) (models.V5GetNodePoolsResponse, error) {
+	clientV2, err := client.NewWithConfig(flags.CmdAPIEndpoint, flags.CmdToken)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	auxParams := clientV2.DefaultAuxiliaryParams()
+	auxParams.ActivityName = activityName
+
+	response, err := clientV2.GetNodePools(args.clusterID, auxParams)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// sort node pools by ID
+	sort.Slice(response.Payload[:], func(i, j int) bool {
+		return response.Payload[i].ID < response.Payload[j].ID
+	})
+
+	return response.Payload, nil
+
+}
+
+func printResult(cmd *cobra.Command, positionalArgs []string) {
+	args := defaultArgs(positionalArgs)
+	nodePools, err := fetchNodePools(args)
+	if err != nil {
+		errors.HandleCommonErrors(err)
+		client.HandleErrors(err)
+
+		if clientErr, ok := err.(*clienterror.APIError); ok {
+			fmt.Println(color.RedString(clientErr.ErrorMessage))
+			if clientErr.ErrorDetails != "" {
+				fmt.Println(clientErr.ErrorDetails)
+			}
+		} else {
+			fmt.Println(color.RedString("Error: %s", err.Error()))
+		}
+		os.Exit(1)
+	}
+
+	awsInfo, err := nodespec.NewAWS()
+	if err != nil {
+		fmt.Println(color.RedString("Error: Cannot provide info on AWS instance types. Details: %s", err))
```

Comment:
As a general rule of thumb, errors should always be propagated back to the caller, so they can be handled there if desired. A program flow should be simple. A program starts, a program ends. A program should end where it started. This here now prints something and still execution will proceed. From here more will go wrong probably.

Reply:
The intend here was to report this error, but treat it as non-breaking. Anyway, I think I will change that to avoid this special treatment here.

Reply:
This is changed now. I gathered all data collection in a common function, moving this part out of the `printResult` function.